flow/diff_context.h

```diff
@@ -0,0 +1,235 @@
+#ifndef FLUTTER_FLOW_DIFF_CONTEXT_H_
+#define FLUTTER_FLOW_DIFF_CONTEXT_H_
+
+#include <map>
+#include <vector>
+#include "flutter/fml/logging.h"
+#include "flutter/fml/macros.h"
+#include "third_party/skia/include/core/SkMatrix.h"
+#include "third_party/skia/include/core/SkRect.h"
+
+namespace flutter {
+
+#ifndef NDEBUG
+#define FLUTTER_ENABLE_DIFF_CONTEXT
+#endif
+
+#ifdef FLUTTER_ENABLE_DIFF_CONTEXT
+
+class Layer;
+
+struct Damage {
+  // This is the damage between current and previous frame;
+  // If embedder supports partial update, this is the region that needs to be
+  // swapped.
+  SkIRect surface_damage;
+
+  // Reflects actual change to target framebuffer; This is surface_damage +
+  // damage previously acumulated for target framebuffer.
+  // All drawing will be clipped to this region. Knowing the affected area
+  // upfront may be useful for tile based GPUs.
+  SkIRect buffer_damage;
```

Comment:
The names `surface_damage` and `buffer_damage` were chosen to be consistent with [EGL_KHR_partial_update](https://www.khronos.org/registry/EGL/extensions/KHR/EGL_KHR_partial_update.txt) terminology. I do remember mentioning it in the code but perhaps it was in the latter PRs. I'll have to look into this.

The area needed to be swapped (current_buffer - front_buffer) is always subset of buffer damage. That's because front buffer is updated on every frame, where as current buffer lags two or three frames behind. 

So for double buffering, to get current buffer up-to-date, you need to repaint area (N, N-1) and (N-1, N-2), but (N-1, n-2) has already been swapped in previous frame, so you only need to swap (N, N-1).

Analogous for triple buffering, to get current buffer up-to-date (assuming buffers are repeated in consistent A, B, C pattern), to get current buffer up-to-date, you need to repaint areas (N, N-1), (N-1, N-2) and (N-2, N-3). But the (N-1,N-2) and (N-2,N-3) parts of screen already show proper contents (they have been updated in last two frames), so only part (N, N-1) needs to be swapped.

Reply:
I get the concepts, but the terminology is unfortunately in conflict with the code base in which it is appearing.

In Skia parlance and in the Flutter engine naming scheme, Surface == Buffer. A Skia surface manages pixels for the destination for a canvas and we have such an SkSurface (and a flutter::Surface) associated with each back buffer.

Perhaps `frame_damage` or `layer_tree_damage` for the N vs N-1 damage and `surface_damage` or `buffer_damage` for the accumulated damage? Then associate them with the terms used in the extension in the comments.

Reply:
I like frame_damage and buffer_damage. Fixed [here](https://github.com/flutter/engine/pull/21824/commits/d57581f4e5a29dd0a49ba49417adae4041fca6b3).